tests/helper/parameter_checks.hpp

```diff
@@ -0,0 +1,146 @@
+      }
+    }
+
+    template <typename ParamT, typename ElementT, typename ValueT,
+              typename ComparedT>
+    void check_default_impl(ElementT element, ValueT modifiedVal,
+                            ComparedT expectedDefault,
+                            ComparedT expectedAfterSet) {
+      using boost::typeindex::type_id;
+      INFO("element type: " << type_id<ElementT>().pretty_name());
+      INFO("parameter type: " << type_id<ParamT>().pretty_name());
+      SECTION("check default constructed element") {
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == true);
+        REQUIRE(element.template get<ParamT>() == expectedDefault);
+      }
+      SECTION("check parameter after set") {
+        element.set(ParamT(modifiedVal));
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+        REQUIRE(element.template get<ParamT>() == expectedAfterSet);
+        SECTION("and unset") {
+          element.template unset<ParamT>();
+          REQUIRE(element.template has<ParamT>() == true);
+          REQUIRE(element.template isDefault<ParamT>() == true);
+          REQUIRE(element.template get<ParamT>() == expectedDefault);
+        }
+      }
+    }
+
+    template <typename ParamT, typename ElementT, typename ValueT,
+              typename ComparedT>
+    void check_optional_impl(ElementT element, ValueT modifiedVal,
+                             ComparedT expectedAfterSet) {
+      using boost::typeindex::type_id;
+      INFO("element type: " << type_id<ElementT>().pretty_name());
+      INFO("parameter type: " << type_id<ParamT>().pretty_name());
+      SECTION("check default constructed element") {
+        REQUIRE(element.template has<ParamT>() == false);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+      }
+      SECTION("check parameter after set") {
+        element.set(ParamT{modifiedVal});
+        REQUIRE(element.template has<ParamT>() == true);
+        REQUIRE(element.template isDefault<ParamT>() == false);
+        REQUIRE(element.template get<ParamT>() == expectedAfterSet);
+        SECTION("and unset") {
+          element.template unset<ParamT>();
+          REQUIRE(element.template has<ParamT>() == false);
+          REQUIRE(element.template isDefault<ParamT>() == false);
+        }
+      }
+    }
+    template <typename ValueT,
+              typename std::enable_if<!std::is_floating_point<ValueT>::value,
+                                      bool>::type = true>
+    ValueT wrapFloat(ValueT value) {
+      return value;
+    }
+
+    template <typename ValueT,
+              typename std::enable_if<std::is_floating_point<ValueT>::value,
+                                      bool>::type = true>
+    Approx wrapFloat(ValueT value) {
+      return Approx(value);
+    }
+  }  // namespace detail
+
+  template <typename T>
+  class HasDefaultOf {
```

Comment:
could these go in the detail namespace too?

Reply:
Oops. They were meant to be. Thanks!